src/main/java/org/commonjava/indy/service/archive/controller/ArchiveController.java

```diff
@@ -130,27 +155,57 @@ public void destroy()
 
     public void generate( HistoricalContentDTO content )
     {
-        int threads = 4 * Runtime.getRuntime().availableProcessors();
-        ExecutorService generateExecutor = Executors.newFixedThreadPool( threads, ( final Runnable r ) -> {
-            final Thread t = new Thread( r );
-            t.setName( "Generate-" + t.getName() );
-            t.setDaemon( true );
-            return t;
-        } );
-        generateExecutor.execute( () -> doGenerate( content ) );
+        String buildConfigId = content.getBuildConfigId();
+        Object lock = buildConfigLocks.computeIfAbsent( buildConfigId, k -> new Object() );
+        synchronized ( lock )
+        {
+            while ( isInProgress( buildConfigId ) )
+            {
+                logger.info( "There is already generation process in progress for buildConfigId {}, try lock wait.",
```

Comment:
can we just throw warning to user that there is already process in progress, please try later ?  sorry if I miss some requirements. ; -)

Reply:
@sswguo You mean don't let the coming process wait, just warn to user and return? Not sure whether PNC care the response of archive generation since this should be at the end of build, if that's the point I will comment to confirm.

Reply:
We can confirm that which case will send two same archive requests in a short time.

Reply:
@sswguo probably large build with large archive file, since PNC UI doesn't allow concurrent builds with the same config ID, the previous build might be already done but generation with download was still in progress, then another build was in progress and use the same workspace to process download or generation, so we need to know if they want to handle the generation response themselves or just let archive go with the right thing.